### ISSUE TITLE
ImageIter is very slow when preprocessing cifar10 in windows

### ISSUE THREAD

**Issue by traveller59** [apache/incubator-mxnet#8113]
## Environment info
Operating System:Windows 10
Package used (Python/R/Scala/Julia):Python
MXNet version:0.11(20170930 from [this](https://github.com/yajiedesign/mxnet/releases))
Python version and distribution:Python 3.6.2 Anaconda 4.4.0
CPU: Core i5 6500
GPU: 1080Ti
### ImageIter is very slow when preprocessing cifar10, multi-thread image preprocessing seems not work.
I want to training a simple resnet model from this [page](https://github.com/tensorflow/models/tree/master/research/resnet), here is the mxnet data loading and preprocessing code:
```Python
os.environ["MXNET_CPU_WORKER_NTHREADS"] = "4"
path = 'E:/Projects/deeplearning/datasets/images/cifar10_mxnet'
data = mx.sym.var('data', shape=[0, 3, 32, 32])
label = mx.sym.var('label', shape=[0])
train_iter = mx.image.ImageIter(128, (3, 32, 32), 
                               path_imgrec =path + '/cifar10/cifar/train.rec',
                               path_imglist=path + '/cifar10/cifar/train.lst',
                               rand_crop=True, rand_mirror=True, std=np.array([255.0, 255.0, 255.0]),brightness=63. / 255.,
                                contrast=0.8,
                                saturation=0.5,
                              data_name=data.name, label_name=label.name)
val_iter = mx.image.ImageIter(128, (3, 32, 32), 
                               path_imgrec =path + '/cifar10/cifar/test.rec',
                               path_imglist=path + '/cifar10/cifar/test.lst',
                               rand_crop=False, rand_mirror=False, std=np.array([255.0, 255.0, 255.0]),
                              data_name=data.name, label_name=label.name)
```
and training code(model definition code is too long to list here):
```Python
mod = mx.mod.Module(
    symbol=cost,
    context=mx.gpu(),
    # data_names=[d[0] for d in train_iter.data],
    # label_names=[l[0] for l in train_iter.label]
    data_names=[data.name],
    label_names=[label.name])
opt = mx.optimizer.Adam(0.001, wd=2e-4, rescale_grad=1.0/128.0)
mod.fit(train_iter,
        eval_data=val_iter,
        optimizer=opt,
        eval_metric=['ce', 'acc'],
        num_epoch=2,
        # batch_end_callback=mx.callback.Speedometer(128, 200),
        epoch_end_callback=mx.callback.do_checkpoint(save_path + '/model'),
        )
```
When training without ImageIter and preprocessing(using NDArrayIter), mxnet code is 2x faster than my tensorflow code. But when add above image preprocessing, mxnet is alomst 2x slower than tensorflow(tensorflow using ndarray iter and same preprocessing tensor, not read from file). Setting MXNET_CPU_WORKER_NTHREADS from 1 to 4 leads to same training performance, and the GPU load is only 20%(tensorflow is 50%), CPU load is 58%, mechanical hard disk load is less than 10% and sometime is zero.

**Comment by yajiedesign** [comment]
try use mx.io.ImageRecordIter

**Comment by traveller59** [comment]
@yajiedesign Thanks for reply, ImageRecordIter is fast and effective, but the iter still use one thread even if I set "preprocess_threads" to 4:
```
[19:15:22] D:\Program Files (x86)\Jenkins\workspace\mxnet\mxnet\src\io\iter_image_recordio_2.cc:153: ImageRecordIOParser2: E:/Projects/deeplearning/datasets/images/cifar10_mxnet/cifar10/cifar/train.rec, use 1 threads for decoding..
```
Then I check the mxnet source code, and find following code:
```C++
// src/io/iter_image_recordio_2.cc
  #pragma omp parallel
  {
    // be conservative, set number of real cores
    maxthread = std::max(omp_get_num_procs() / 2 - 1, 1);
  }
param_.preprocess_threads = std::min(maxthread, param_.preprocess_threads);
  #pragma omp parallel num_threads(param_.preprocess_threads)
  {
    threadget = omp_get_num_threads();
  }
  param_.preprocess_threads = threadget;
  if (param_.verbose) {
    LOG(INFO) << "ImageRecordIOParser2: " << param_.path_imgrec
              << ", use " << threadget << " threads for decoding..";
  }
// include/dmlc/omp.h
inline int omp_get_num_procs() __GOMP_NOTHROW { return 1; }
inline int omp_get_max_threads() __GOMP_NOTHROW { return 1; }
```
These code indicates that ImageRecordIter always use only one thread, "preprocess_threads" argument in ImageRecordIter is no use.

**Comment by ptrendx** [comment]
That is not true - you have either dual core CPU or 4-core without hyperthreading if you get only 1 thread for decoding. The idea here is to limit the number of OpenMP threads to the number of physical cores - 1, so that the scheduler is not starved for resources, even if preprocess_threads variable is higher than that. We could lift that restriction and rely on user to provide the option that makes sense.

**Comment by yajiedesign** [comment]
`// include/dmlc/omp.h
inline int omp_get_num_procs() __GOMP_NOTHROW { return 1; }
inline int omp_get_max_threads() __GOMP_NOTHROW { return 1; }`
only use when not openmp.but the my relase is use openmp.
and what you cpu core?

**Comment by traveller59** [comment]
Thanks! Now the only problem is ImageIter's performance(ImageRecordIter has no problem). maybe someone can consider to fix it.
